src/HtmlProcessor/HtmlParser.php

```diff
@@ -0,0 +1,208 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pelago\Emogrifier\HtmlProcessor;
+
+/**
+ * Parse a HTML document fragment and add missing root elements.
```

Comment:
Is this a class you've used elsewhere?  I'd like more information on why it is needed, and what deficiencies in the HTML5 library it is covering.

Reply:
I wrote it but it's not yet been released to any customers. It's needed to handle https://github.com/Masterminds/html5-php/issues/166. You can see two different implementations here:
* https://github.com/roundcube/roundcubemail/blob/master/program/lib/Roundcube/rcube_washtml.php#L842
* https://github.com/ezyang/htmlpurifier/blob/master/library/HTMLPurifier/Lexer/DOMLex.php#L310

It's not strictly required in the context of this lib if you're happy for it to barf on malformed HTML

Reply:
> It's needed to handle [Masterminds/html5-php#166](https://github.com/Masterminds/html5-php/issues/166). You can see two different implementations here:
> 
>     * https://github.com/roundcube/roundcubemail/blob/master/program/lib/Roundcube/rcube_washtml.php#L842
> 
>     * https://github.com/ezyang/htmlpurifier/blob/master/library/HTMLPurifier/Lexer/DOMLex.php#L310

Thanks.  I think I understand better now: `masterminds/html5-php` treats input as a document fragment rather than a complete document (even if it has a `DOCTYPE`??) and does not (currently) adhere the W3C spec for `<html>`. `<head>` and perhaps `<body>` elements (linked from the discussion in Masterminds/html5-php#166).

It would be good to put all this in the PHPDoc for the class.  (Now I read again what you have put, with the benefit of having looked at the discussion, I see that it's a nice succinct summary 😊 but it wasn't immediately obvious what it was all about.)

> It's not strictly required in the context of this lib if you're happy for it to barf on malformed HTML

Adding missing `DOCTYPE`, `<html>`, `<head>` and `<body>` elements has for a long time been part of Emogrifier.  I think we would end up breaking people's code without them realizing if we stopped doing this.

(This does make me wonder, however, if `Masterminds/html5-php` actually deals properly with all the rules of HTML5 where closing tags can be completely omitted depending what element is next or what the parent is.  E.g. as a simple example, `<div><p>First pargraph in 1st DIV.<p>Second paragraph in 1st DIV<div>Sibling of 1st DIV` is (I think) perfectly valid HTML [though I suspect `DOMDocument` doesn't deal with this kind of thing very well either, particularly as it predates HTML5 AFAIK].)

Reply:
This class has been released in beta and will soon be released in stable so I'll correct if anything comes up.

> It would be good to put all this in the PHPDoc for the class.

If you have a suggestion of what you'd like to see I'm happy to add it.

> Adding missing DOCTYPE, <html>, <head> and <body> elements has for a long time been part of Emogrifier.

This would suggest you have some code elsewhere in the lib already that achives the same as this class?

Reply:
> > It would be good to put all this in the PHPDoc for the class.
> 
> If you have a suggestion of what you'd like to see I'm happy to add it.

Probably to clarify which missing root elements are added (is it `<html>`, `<head>` and `<body>` though not `<!DOCTYPE>`?), and that this is because `masterminds/html5` treats the input HTML as a document fragment rather than complete document (even if it has a DOCTYPE), so these elements do not get created if their tags are not in the provided HTML
.
> > Adding missing DOCTYPE, ,  and  elements has for a long time been part of Emogrifier.
> 
> This would suggest you have some code elsewhere in the lib already that achives the same as this class?

`AbstractHtmlProcessor::ensureDocumentType` and `AbstractHtmlProcessor::addContentTypeMetaTag` ensure a DOCTYPE and `Content-Type` are present.  The creation of the `<html>`, `<head>` and `<body>` elements (whether or not explicitly provided in the input HTML) is done by `DOMDocument` itself, so there is no specific code in Emogrifier to do this, though it is tested that these elements are present in the output HTML.